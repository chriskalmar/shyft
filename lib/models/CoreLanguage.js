
import {
  dataTypes,
  Entity,
} from '../../';


const {
  DataTypeString,
} = dataTypes



export const CoreLanguage = new Entity({
  name: 'language',
  domain: 'core',
  description: 'A language',

  // indexing: {
  //   unique: [
  //     [ 'name' ],
  //     [ 'iso_code' ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      maxLength: 30,
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      maxLength: 30,
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      pattern: '^[a-z]+$',
      minLength: 2,
      maxLength: 2,
      required: true
    },
  }

})

