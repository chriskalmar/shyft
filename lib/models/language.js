
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'language',
      description: 'a language',

      cascadeOnDelete: true,

      attributes: [

        {
          name: 'name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'native_name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'iso_code',
          type: 'string',
          description: 'ISO 639-1 code of the language',
          maxLength: 2,
          required: true
        },

      ],

      indexing: {
        unique: [
          [ 'name' ],
          [ 'iso_code' ],
        ]
      }

    }

  ]

}
