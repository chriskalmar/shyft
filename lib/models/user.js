
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'user',
      description: 'a user',

      isUserTable: true,
      cascadeOnDelete: true,

      attributes: [

        {
          name: 'username',
          type: 'string',
          description: 'Username',
          pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
          minLength: 3,
          maxLength: 30,
          required: true
        },

        {
          name: 'first_name',
          type: 'string',
          description: 'First name',
          maxLength: 30
        },

        {
          name: 'last_name',
          type: 'string',
          description: 'Last name',
          maxLength: 30
        },

        {
          name: 'language',
          type: 'reference',
          description: 'User language',
          target: 'core::language',
          required: true
        },

        {
          name: 'email',
          type: 'email',
          description: 'Email address',
          required: true
        },

        {
          name: 'password',
          type: 'password',
          description: 'User password',
          required: true
        }
      ],

      indexing: {
        unique: [
          [ 'username' ],
          [ 'email' ],
        ]
      }

    }

  ]

}
