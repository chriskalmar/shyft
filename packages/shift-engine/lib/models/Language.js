
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';


export const Language = new Entity({
  name: 'Language',
  description: 'A language',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'name' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'isoCode' ],
    }),
  ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      required: true
    },
  }

})

