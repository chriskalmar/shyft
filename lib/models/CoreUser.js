
import {
  datatypes,
  Entity,
} from '../../';

import CoreLanguage from './CoreLanguage';

const {
  DataTypeString,
} = datatypes

export const CoreUser = new Entity({
  name: 'user',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  // indexing: {
  //   unique: [
  //     [ 'username' ],
  //     [ 'email' ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: CoreLanguage,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



