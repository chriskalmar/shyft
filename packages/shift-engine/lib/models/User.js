
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';


export const User = new Entity({
  name: 'User',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexing: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: Language,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



