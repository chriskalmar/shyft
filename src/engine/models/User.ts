import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
