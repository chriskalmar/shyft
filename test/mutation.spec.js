import './setupAndTearDown';
import {
  mutate,
} from './db';

import {
  asUser,
  removeDynamicData,
} from './testUtils';

import { BoardMember } from './models/BoardMember';
import { Board } from './models/Board';
import { Book } from './models/Book';


describe('mutation', () => {

  it('reject mutations without state transitions on stateful entities', async () => {

    const payload = {
      board: 50,
    }

    await mutate(BoardMember, 'create', payload, null, asUser(99))
      .catch(e => {
        expect(e).toMatchSnapshot()
      })

  })



  it('perform create mutations', async () => {

    const payload = {
      board: 47,
    }

    const result = await mutate(BoardMember, 'join', payload, null, asUser(99))
    expect(removeDynamicData(BoardMember, result)).toMatchSnapshot()
  })


  it('perform create mutations with nested JSON attributes', async () => {

    const payload = {
      title: 'War and Peace',
      author: 'Leo Tolstoy',
      reviews: [
        {
          reviewer: 'John Connor',
          reviewText: 'Couldn\'t stop reading',
          bookAttributes: [
            {
              attribute: 'Year of publishing',
              value: '1867',
            },
            {
              attribute: 'Pages',
              value: '1225',
            },
          ]
        },
      ]
    }

    const result = await mutate(Book, 'create', payload, null, asUser(99))
    expect(result).toMatchSnapshot()
  })


  it('perform update mutations', async () => {

    const payload = {
      board: 50,
      invitee: 80,
    }

    let result = await mutate(BoardMember, 'invite', payload, null, asUser(84))
    const inviteId = result.id
    expect(removeDynamicData(BoardMember, result)).toMatchSnapshot()

    result = await mutate(BoardMember, 'accept', {}, inviteId, asUser(80))
    expect(removeDynamicData(BoardMember, result)).toMatchSnapshot()
  })


  it('handle uniqueness constraints', async () => {

    const payload = {
      name: 'New Board',
      isPrivate: false,
    }

    await mutate(Board, 'build', payload, null, asUser(99))

    await mutate(Board, 'build', payload, null, asUser(99))
      .catch(e => {
        expect(e).toMatchSnapshot()
      })
  })


  it('perform delete mutations', async () => {

    const payload = {
      board: 50,
      invitee: 81,
    }

    let result = await mutate(BoardMember, 'invite', payload, null, asUser(84))
    const inviteId = result.id
    expect(removeDynamicData(BoardMember, result)).toMatchSnapshot()

    result = await mutate(BoardMember, 'remove', {}, inviteId, asUser(84))
    expect(removeDynamicData(BoardMember, result)).toMatchSnapshot()
  })


  it('fail mutation if fromState is not a match', async () => {

    const payload = {
      board: 50,
      invitee: 81,
    }

    const result = await mutate(BoardMember, 'invite', payload, null, asUser(84))
    const inviteId = result.id

    await mutate(BoardMember, 'accept', {}, inviteId, asUser(81))

    await mutate(BoardMember, 'accept', {}, inviteId, asUser(81))
      .catch(e => {
        expect(e).toMatchSnapshot()
      })
  })



  it('reject mutations if ID not found', async () => {

    await mutate(BoardMember, 'remove', {}, 99999, asUser(80))
      .catch(e => {
        expect(e).toMatchSnapshot()
      })
  })


  it('reject unknown mutations', async () => {

    await mutate(BoardMember, 'findInnerPeace', {}, 1, asUser(80))
      .catch(e => {
        expect(e).toMatchSnapshot()
      })
  })

})
