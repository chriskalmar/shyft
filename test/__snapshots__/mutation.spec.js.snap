// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation fail mutation if fromState is not a match 1`] = `[Error: Instance with ID '304' not found or access denied]`;

exports[`mutation handle uniqueness constraints 1`] = `[Error: Uniqueness constraint violated]`;

exports[`mutation perform create mutations 1`] = `
Object {
  "board": "47",
  "id": "301",
  "invitee": "99",
  "inviter": "99",
  "state": 50,
}
`;

exports[`mutation perform create mutations with nested JSON attributes 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "id": "1",
  "reviews": Array [
    Object {
      "bookAttributes": Array [
        Object {
          "attribute": "Year of publishing",
          "value": "1867",
        },
        Object {
          "attribute": "Pages",
          "value": "1225",
        },
      ],
      "reviewText": "Couldn't stop reading",
      "reviewer": "John Connor",
    },
  ],
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "default": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": undefined,
    "default": "War and Peace",
  },
}
`;

exports[`mutation perform delete mutations 1`] = `
Object {
  "board": "50",
  "id": "303",
  "invitee": "81",
  "inviter": "84",
  "state": 10,
}
`;

exports[`mutation perform delete mutations 2`] = `
Object {
  "deleteRowCount": 1,
  "id": "303",
}
`;

exports[`mutation perform update mutations 1`] = `
Object {
  "board": "50",
  "id": "302",
  "invitee": "80",
  "inviter": "84",
  "state": 10,
}
`;

exports[`mutation perform update mutations 2`] = `
Object {
  "board": "50",
  "id": "302",
  "invitee": "80",
  "inviter": "84",
  "state": 20,
}
`;

exports[`mutation reject mutations if ID not found 1`] = `[Error: Instance with ID '99999' not found or access denied]`;

exports[`mutation reject mutations without state transitions on stateful entities 1`] = `[Error: State transition (toState) not defined]`;

exports[`mutation reject unknown mutations 1`] = `[Error: Invalid mutation]`;
